src/pg_query_summary.c

```diff
@@ -0,0 +1,761 @@
+#include "pg_query.h"
+#include "pg_query_internal.h"
+#include "pg_query_outfuncs.h"
+
+#include "parser/parser.h"
+#include "parser/scanner.h"
+#include "parser/scansup.h"
+
+#include "nodes/nodeFuncs.h"
+#include "utils/builtins.h"
+
+#include "protobuf/pg_query.pb-c.h"
+
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include "pg_query_summary.h"
+
+static void add_table(WalkState* state, ContextType context, char* schemaname, char* relname);
+
+static PgQueryError* make_pg_query_error(const char* message)
+{
+	PgQueryError* error = palloc(sizeof(PgQueryError));
+	error->message = pstrdup(message);
+	return error;
+}
+
+// Given a Node (node) and WalkState (state), wraps node in a NodeWithContext
+// and adds it to the state->from_clause_items list.
+static void add_from_clause_item(void *node, ContextType context, WalkState *state) {
+	if (!node)
+		return;
+
+	NodeWithContext* nwc = palloc(sizeof(NodeWithContext));
+	nwc->node = node;
+	nwc->context = context;
+
+	state->from_clause_items = lappend(state->from_clause_items, nwc);
+}
+
+// Given a List*-of-Nodes (clause) and WalkState (state), wrap them all
+// in NodeWithContext structs and add them to the state->from_clause_items list.
+static void add_from_clause_items(List *clause, ContextType context, WalkState *state) {
+	if (!clause)
+		return;
+
+	ListCell *lc = NULL;
+	foreach(lc, clause) {
+		Node* node = lfirst(lc);
+		add_from_clause_item(node, context, state);
+	}
+}
+
+static void add_function(List* funcname, ContextType context, WalkState* state){
+	StringInfoData buf;
+	initStringInfo(&buf);
+
+	ListCell* lc = NULL;
+	foreach(lc, funcname) {
+		if (list_cell_number(funcname, lc) != 0)
+			appendStringInfoChar(&buf, '.');
+
+		String* str = lfirst_node(String, lc);
+		appendStringInfoString(&buf, quote_identifier(str->sval));
+	}
+
+	SummaryFunction* func = palloc(sizeof(SummaryFunction));
+
+	func->name = buf.data;
```

Comment:
We can simplify this with [NameListToString](https://github.com/postgres/postgres/blob/master/src/backend/catalog/namespace.c#L3594):
```suggestion
	SummaryFunction* func = palloc(sizeof(SummaryFunction));

	func->name = NameListToString(funcname);
```